design/external-references.md

```diff
@@ -0,0 +1,394 @@
+
+As mentioned about, the default resolve type for imports is `Synthesis` (with strong-references). This means that the values returned by `readString` will be actual concrete values. If users opt-in to deploy-time resolution (by setting using `Deployment` resolve type), the values returned will be tokens. In some cases this would be fine, but there could be constructs that cannot deal with opaque values (i.e. if the value is an list of strings and needs to be deconstructed). In those cases (REQ8), constructs should invoke `readString` with `allowUnresolved: false` to indicate that this specific value cannot be a token.
+
+#### Export names
+
+AWS CloudFormation export names must be unique within an environment (account/region), and they will be formed by concatenating root `exportName` and all the keys that lead to a value in the serialization tree.
+
+We will use `-` as a component separator devising fully qualified export names. To avoid collisions, if the main export name or any subsequent serialization key includes a `-` it will be removed.
+
+Since AWS CloudFormation has a limit on export name length, and we wouldn't want to restrict the serialization depth, the import/export serializer should trim the name and add a hash of the full name, but only if the total length exceeds the limit.
+
+### Synthesis-time Imports (REQ8)
+
+As mentioned in the previous section, `importObject` will support both synthesis and deploy-time imports by export name.
+
+In order to implement synthesis-time imports, we will add a new environmental context provider to the toolkit which will be able to retrieve a value for a certain CloudFormation named export.
+
+The implementation of this provider will use the CloudFormation ListExports operation to find the exports needed and pass their values in through the CDK context mechanism.
+
+Bear in mind that once a value has been retrieved, it will automatically be saved in the local `cdk.json` and won't be retrieved again until `cdk context --reset` is called.
+
+Since synthesis-time resolution doesn't create strong coupling between the stacks at the CloudFormation
+level, production and operational issues can arise if the producing stack deletes an exported resource. On the other hand, we hear from customers that strong-referencing behavior of `Fn::ImportValue` is sometimes a curse. Customers tell us that they found themselves stuck with unremovable or stacks that cannot be updates due to imports.
+
+Luckily, we can enable both capabilities. By default, when synthesis-time resolution is used, the CDK will automatically add a resource metadata entry to the template with an `Fn::ImportValue`. This will create the strong coupling between the stacks. Users can opt-out of this behavior by setting `weak: true` when they import the resource.
```

Comment:
We could also emulate the `Fn::Import` lock by automatically adjusting the stack policy to prevent deletion/replaces of the exported resources.